generator/PythonGenerator.py

```diff
@@ -0,0 +1,622 @@
+"""
+Generates openMINDS classes.
+
+"""
+
+import glob
+import json
+import os
+import re
+from typing import List
+from collections import defaultdict
+import warnings
+
+from generator.commons import (
+    JinjaGenerator,
+    TEMPLATE_PROPERTY_TYPE,
+    TEMPLATE_PROPERTY_LINKED_TYPES,
+    SchemaStructure,
+    TEMPLATE_PROPERTY_EMBEDDED_TYPES,
+    SCHEMA_FILE_ENDING,
+    ROOT_PATH,
+    EXPANDED_DIR,
+    find_resource_directories)
+
+LIST_CLASSES_TEMPATE = '''
+
+def list_kg_classes():
+    """List all KG classes defined in this module"""
+    return [obj for name, obj in inspect.getmembers(sys.modules[__name__])
+           if inspect.isclass(obj) and issubclass(obj, KGObject) and obj.__module__.startswith(__name__)]
+'''
+
+# for backwards compatibility or to increase clarity we remap certain
+# names from the schemas when creating Python attribute names
+name_map = {
+    "shortName": "alias",
+    "fullName": "name",
+    "scope": "model_scope",
+    "hasVersion": "versions",
+    "hasEntity": "entities",
+}
+
+# in general we make attribute names plural when the attribute can contain multiple items
+# the following dict contains exceptions to the simple rule we use for making names plural
+# (i.e. add 's' unless the word already ends in 's')
+custom_multiple = {
+    "data": "data",
+    "input_data": "input_data",
+    "output_data": "output_data",
+    "reference_data": "reference_data",
+    "funding": "funding",
+    "biological_sex": "biological_sex",
+    "age_category": "age_categories",
+    "descended_from": "descended_from",
+    "laterality": "laterality",
+    "software": "software",
+    "configuration": "configuration",
+    "is_part_of": "is_part_of",
+    "semantically_anchored_to": "semantically_anchored_to",
+    "is_alternative_version_of": "is_alternative_version_of",
+    "experimental_approach": "experimental_approaches",
+    "pathology": "pathologies",
+    "uncertainty": "uncertainties",
+    "application_category": "application_categories",
+    "about": "about"
+}
+
+
+def generate_python_name(json_name, allow_multiple=False):
+    if json_name in name_map:
+        python_name = name_map[json_name]
+    else:
+        python_name = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", json_name)
+        python_name = re.sub("([a-z0-9])([A-Z])", r"\1_\2", python_name).lower()
+        if allow_multiple and python_name[-1] != "s":
```

Comment:
This point should be discussed. fairgraph uses plural names to refer to properties that _can_ be arrays, as I feel this is more Pythonic.

In openMINDS property names, plural names are only used if the values _must_ be arrays. I don't like this rule, but I can live with it.

We should decide what to do for openMINDS Python. I would prefer to go for what I see as the more Pythonic approach, but there are arguments to stay more consistent with the openMINDS schemas.

Reply:
I discussed this with @lzehl and it makes sense, for consistency, to keep everything singular.